s3box_review/s3box.go

```diff
@@ -0,0 +1,243 @@
+	if options.S3Bucket == "" {
+		return nil, ErrS3BucketRequired
+	}
+
+	bufferSize := DefaultBufferSize
+	if options.BufferSize > 0 {
+		bufferSize = options.BufferSize
+	}
+
+	// Setup s3 handler and aws configuration. If no creds are explicitly provided, they'll be grabbed from the environment.
+	region, err := getRegionForBucket(options.S3Bucket)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get AWS region for bucket %s: (%s)", options.S3Bucket, err)
+	}
+
+	// If AWS creds were provided use those, otherwise grab them from your environment
+	var awsCreds *credentials.Credentials
+	if options.AWSKey == "" && options.AWSPassword == "" && options.AWSToken == "" {
+		awsCreds = credentials.NewEnvCredentials()
+	} else {
+		awsCreds = credentials.NewStaticCredentials(options.AWSKey, options.AWSPassword, options.AWSToken)
+	}
+	awsConfig := aws.NewConfig().WithRegion(region).WithS3ForcePathStyle(true).WithCredentials(awsCreds)
+	awsSession := session.New()
+
+	return &S3Box{
+		s3Bucket:   options.S3Bucket,
+		timestamp:  time.Now(),
+		s3Handler:  s3.New(awsSession, awsConfig),
+		bufferSize: bufferSize,
+	}, nil
+}
+
+// NextBox gives you a new box, forgetting everything about previously packaged data
+func (sb *S3Box) NextBox() {
+	sb.Lock()
+	sb.timestamp = time.Now()
+	sb.fileNumber = 0
+	sb.fileLocations = []string{}
+	sb.bufferedData = []byte{}
+	sb.isSealed = false
+	sb.Unlock()
+}
+
+// Pack writes bytes into a buffer. Once that buffer hits capacity, the data is output to s3.
+// Any error will leave the buffer unmodified.
+func (sb *S3Box) Pack(data []byte) error {
+	if sb.isSealed {
+		return ErrBoxIsSealed
+	}
+
+	// If the bytes aren't in JSON format, return an error
+	var tempMap map[string]interface{}
+	if err := json.Unmarshal(data, &tempMap); err != nil {
+		return ErrInvalidJSONInput
+	}
+
+	sb.Lock()
+	oldBuffer := sb.bufferedData // If write fails, keep buffered data unchanged
+	data = append(data, '\n')    // Append a new line for text-editor readability
+	sb.bufferedData = append(sb.bufferedData, data...)
+	sb.Unlock()
+
+	// If we're hitting capacity, dump the results to s3.
+	// If shipping to s3 errors, don't modify the buffer.
+	if len(sb.bufferedData) > sb.bufferSize {
+		if err := sb.dumpToS3(); err != nil {
+			sb.Lock()
+			sb.bufferedData = oldBuffer
+			sb.Unlock()
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Seal closes writes and flushes any buffered data to s3.
+// Once the box is sealed, you can no longer write to this box.
+// Use NextBox to create a new box if you'd like to continue packing.
+func (sb *S3Box) Seal() error {
+	if err := sb.dumpToS3(); err != nil {
+		return err
+	}
+
+	sb.isSealed = true
+	return nil
+}
+
+// dumpToS3 ships buffered  data to s3 and increments the index with a clean slate of running data
+func (sb *S3Box) dumpToS3() error {
+	if len(sb.bufferedData) == 0 {
+		return nil
+	}
+	fileKey := fmt.Sprintf("%d_%d.json.gz", sb.timestamp.UnixNano(), sb.fileNumber)
+	sb.Lock()
```

Comment:
Thinking of moving the lock above. If `Seal` is called twice in quick succession, the first call will upload the remaining data, while the second call will attempt to write nothing. This wouldn't be disastrous, but would add noise and messiness to s3.

This being said, the user should _never_ need to `Seal` more than once.